internal/web/client_test.go

```diff
@@ -284,3 +429,79 @@ func createTestServerWithOAuth2(bearer string) *httptest.Server {
 		writer.WriteHeader(http.StatusUnauthorized)
 	}))
 }
+
+// createTestServerWithAPIKey is a helper that creates a httptest.Server with a handler that has api key auth.
+func createTestServerWithAPIKey(key, passphrase, secret string) *httptest.Server {
+	return httptest.NewServer(http.HandlerFunc(func(writer http.ResponseWriter, req *http.Request) {
+		reqContentType := req.Header.Get("content-type")
+		if reqContentType != "application/json" {
+			writer.WriteHeader(http.StatusUnsupportedMediaType)
+
+			return
+		}
+
+		reqKey := req.Header.Get("cb-access-key")
+		reqPassphrase := req.Header.Get("cb-access-passphrase")
+		reqSign := req.Header.Get("cb-access-sign")
+		if reqKey != key || reqPassphrase != passphrase || reqSign == "" {
+			writer.WriteHeader(http.StatusUnauthorized)
+
+			return
+		}
+
+		// generate sign
+		msg := generageMsg(req, req.Header.Get("cb-access-timestamp"))
+
+		sign, err := generateSign(secret, msg)
+		if err != nil {
+			writer.WriteHeader(http.StatusInternalServerError)
+
+			return
+		}
+
+		// compare signatures
+		if reqSign != sign {
+			writer.WriteHeader(http.StatusUnauthorized)
+
+			return
+		}
+
+		writer.WriteHeader(http.StatusOK)
+	}))
+}
+
+// generateSign is a helper to generate signature for request.
+func generateSign(secret, message string) (string, error) {
```

Comment:
@prestonvasquez Can you explain this line to me? It's from [here](https://github.com/alpine-hodler/gidari/blob/main/internal/web/auth/api_key.go#L110). It won't always be an empty string?
```go
postAuthority := strings.Replace(req.URL.String(), auth.url.String(), "", 1)
```

I rewrote it like this because it worked on the test server, but moving this to the `tools` package might not be right.
```go
postAuthority := strings.TrimSuffix(req.URL.String(), "/")
```

Reply:
The "authority" is [this part](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_is_a_URL#authority) of the URL. We want everything after that for the message. Hence "Post Authority". Just Add a new argument. I think you can use the URL path for this:

```go
	return fmt.Sprintf("%s%s%s%s", timestamp, req.Method, req.URL.Path, string(body))
```

When you make the changes I'll check your code out and use an API that requires signed messages to be 100%.

Reply:
I.e., this

```go
// NewHTTPMessage generates the  base64-encoded message required to make API-Key-Authenticated requests.  
func NewHTTPMessage(req *http.Request, timestamp string) HTTPMessage {
	if req.Body == nil {
		return []byte{}
	}

	body, _ := io.ReadAll(req.Body)
	req.Body = io.NopCloser(bytes.NewBuffer(body))
	
	return fmt.Sprintf("%s%s%s%s", timestamp, req.Method, req.URL.Path, string(body))
}
```

Reply:
@prestonvasquez I did, but in a different way, the solution with array gave index out of range.